Add user manual for observe

kitchoi opened the pull request:
This PR add user manual for the `observe` framework.

- The documentation for `on_trait_change` is moved to `listening.rst`. The first paragraph is extracted as an intermediate index page for notifications.
- Add `observing.rst` for the `observe` framework.
- Add `traits.observation.api` to the API reference for cross-referencing.
- Add examples to `examples/tutorials/doc_examples`

Edited (this was the previous message:)
~This is a draft PR for making the `observe` user manual visible (see https://github.com/enthought/traits/issues/977).~

~Much of the API cross-references are missing in this PR, as the PR would otherwise be too big. To see the documentation with styling and cross-referencing, checkout `observer-framework-test` branch and build the documentation using `python etstool.py docs`.~

**Checklist**
- ~Tests~
- [x] Update API reference (`docs/source/traits_api_reference`)
- [x] Update User manual (`docs/source/traits_user_manual`)
- ~Update type annotation hints in `traits-stubs`~

corranwebster commented on docs/source/traits_user_manual/observers.rst:
```diff
@@ -0,0 +1,557 @@
+There are two approaches for creating an expression:
+
+* Expressions as objects
+* Text using Traits domain specific language
```
maybe "Traits observer mini-language" or "Traits observer pattern langauge" or something similar?

corranwebster commented on docs/source/traits_user_manual/observers.rst:
```diff
@@ -0,0 +1,557 @@
+way to specify observation rules via a single text. It supports a subset of
+features in |Expression|.
+
+.. rubric:: Syntax
```
The formal grammar probably doesn't belong here, but it would certainly fit well in an "internals" section, or as part of the main "internals" chapter.

corranwebster commented on docs/source/traits_user_manual/observers.rst:
```diff
@@ -0,0 +1,557 @@
+* Expressions as objects
+* Text using Traits domain specific language
+
+Expressions as objects
```
I'd do the language first, then do the expression object second: start with what 90% of people will probably use, and then give them the advanced way.

corranwebster commented on docs/source/traits_user_manual/listening.rst:
```diff
@@ -0,0 +1,782 @@
+Trait Notification with ``on_trait_change``
+===========================================
+
+Requesting trait attribute change notifications can be done in several
+ways:
+
+.. index:: notification; strategies
+
+* Dynamically, by calling on_trait_change() or on_trait_event() to establish
+  (or remove) change notification handlers.
+* Statically, by decorating methods on the class with the @on_trait_change
+  decorator to indicate that they handle notification for specified attributes.
+* Statically, by using a special naming convention for methods on the class to
+  indicate that they handle notifications for specific trait attributes.
+
+.. index:: notification; dynamic
```
This probably needs to be updated to give the context for the older methods of notification.

corranwebster commented on docs/source/traits_user_manual/observing.rst:
```diff
@@ -0,0 +1,538 @@
+  decorator to indicate that they handle notification for specified attributes.
+* Via instance method: By calling |HasTraits.observe| instance method to establish (or remove) change
+  notification handlers.
+
```
You might mention listeners somewhere here as an older alternative.

corranwebster commented on docs/source/traits_user_manual/observing.rst:
```diff
@@ -0,0 +1,538 @@
+applying the |@observe| decorator on an instance method. It has the following
+signature:
+
+.. py:decorator:: observe(expression[, post_init=False, dispatch="same"])
```
Can this be made to use autodoc so that it doesn't get out of date if we change the API?

kitchoi replied:
autodoc ends up bringing the entire docstring here, which is too verbose, an alternative is to write our custom hooks for `autodoc` but that is probably an overkill. I agree with your concern, so I removed these now and hopefully the links to the API documentation is enough. I reworked these paragraphs to provide the same level of context for the subsequent sections. In particular, the mention of "expression" and "handler" are needed for later sections.

corranwebster commented on docs/source/traits_user_manual/observing.rst:
```diff
@@ -0,0 +1,538 @@
+The |HasTraits.observe| method is useful for adding or removing change handler on a per
+instance basis. The method has the following signature:
+
+.. py:method:: HasTraits.observe(handler, expression[, remove=False, dispatch="same"])
```
Same comment about autodoc here

corranwebster commented on docs/source/traits_user_manual/observing.rst:
```diff
@@ -0,0 +1,538 @@
+Notification Handler
+--------------------
+
+As soon as a change has occurred, the **handler** callable provided will be
```
"As soon as" is vague. We should probably say something about:
- dispatch usually being immediate, but dependent on the `dispatch` keyword (eg. dispatch="ui" is not immediate)
- the order that handlers are called being arbitrary.

corranwebster commented on docs/source/traits_user_manual/observing.rst:
```diff
@@ -0,0 +1,538 @@
+- handler(*event*)
+
+where the *event* parameter is an object representing the change observed.
+The type of *event* depends on the context of the change.
```
We should mention here any guarantees that we make about the common API, eg. if `object` is always present.

corranwebster commented on docs/source/traits_user_manual/observing.rst:
```diff
@@ -0,0 +1,538 @@
+.. rubric:: Notes
+
+#. When a handler is invoked, the change has already occurred.
+#. Multiple handlers can be defined for a given change event and these handlers
```
I'd bump this up, rather than have it in a note (see previous comments).

corranwebster commented on docs/source/traits_user_manual/observing.rst:
```diff
@@ -0,0 +1,538 @@
+   captured and logged.
+
+
+Differences from |@on_trait_change|
```
Maybe "migrating from on_trait_change"?

kitchoi commented:
Thank you @corranwebster, all very good suggestions and I made quite a lot of changes:

- Merged ‘observing’ and ‘notification’ so that observe is the main content for ‘Traits Notification’.
- on_trait_change is moved to an Appendix and is referenced in the introduction for observe.
- Regrouped some of the “differences from on_trait_change” into a section more like a migration guide
- Rebranded some of the “differences from on_trait_change” as features/fixes, with as little mention of on_trait_change as possible.
- Moved the expectations/rules about the change handler up
- Mentioned the “dispatch” parameter and its effect on handler
- Simplified migration-guide-like sections to make them easier to read at a glance.

kitchoi commented on docs/source/traits_user_manual/listening.rst:
```diff
@@ -0,0 +1,786 @@
+introduced for overcoming them. See :ref:`observe-notification` for details and
+for how to migrate.
+
+Change notifications can be set up with **on_trait_change** in several ways:
```
This sentence and the paragraph above are changed. The rest are the same as the original documentation.

kitchoi commented:
I overlooked this important request change:

> the example code should have module docstrings so that the TraitsUI demo code can display it properly

Will continue making changes for this. Removing review request for now. I am sorry for the noise.

mdickinson commented on docs/source/traits_user_manual/notification.rst:
```diff
@@ -9,779 +9,622 @@ possible for trait attributes. This functionality lets you write programs
+
+   * - Expression
+     - Meaning
+   * - *trait("attr1")*
```
Can we add backticks around these, so that the rendered expressions can be copy-pasted? (Currently they're appearing with code-unfriendly smart quotes.)

kitchoi replied:
They will look rather long and squash the "meaning" column into a small column.
I could instead break this out into bulletins instead of trying to fit them in a table.
Happy to do this in a separate PR so not to block the release process.

mdickinson replied:
That could work. Right now, the rendered code can't be copy-and-pasted because of the smart quotes.

kitchoi replied:
SGTM

kitchoi replied:
I thought it was going to take me a while to get the formatting right. It did not take so long at the end: See #1140

mdickinson commented:
This LGTM. There are some places where the wording could be tweaked, but I'm happy for that to happen in a separate PR.

kitchoi commented:
If this is good enough to go in, let's do tweaks in separate PRs so we don't block release preparations.

kitchoi commented:
Thank you both! Merging...
I am working on changes for addressing https://github.com/enthought/traits/pull/1060#discussion_r429142495
That will be in a separate PR.

rahulporuri commented on docs/source/traits_user_manual/notification.rst:
```diff
@@ -9,779 +9,622 @@ possible for trait attributes. This functionality lets you write programs
+     - Meaning
+   * - *attr1\.attr2*
+     - Matches a trait named *attr2* on an object referenced by a trait named
+       *item1* on the current object. Changes to *attr1* or *attr2* will
```
should this be "attr1", not "item1"

kitchoi replied:
Well spotted! Will fix.
